File: src/medium-94/recursion.ts
import { BinaryTreeNode } from 'data-structures/binary-tree'

export function inorderTraversal(node: BinaryTreeNode<number> | null) {
  const answer: number[] = []
  traversal(node, answer)
  return answer
}

function traversal(node: BinaryTreeNode<number> | null, answer: number[]) {
  if (node === null) return

  traversal(node.left, answer)
  answer.push(node.value)
  traversal(node.right, answer)
}

File: src/medium-94/recursion.go
package solution

import structures "leetcode-solutions/data-structures"

// RecursiveInOrderTraversal is medium 94 solution
func RecursiveInOrderTraversal(root *structures.BinaryTreeNode) []int {
	answer := []int{}
	traversal(root, &answer)
	return answer
}

func traversal(root *structures.BinaryTreeNode, answer *[]int) {
	if root == nil {
		return
	}
	traversal(root.Left, answer)
	// type assertion
	// https://tour.golang.org/methods/15
	*answer = append(*answer, root.Val.(int))
	traversal(root.Right, answer)
}
